### PR TITLE
Add a Redis based chat memory store extension

### PR DESCRIPTION
**Pull request by sebastienblanc**
Replace the default `InMemoryChatMemoryStore` by using Redis instead. 
I'm sure if I'm doing the best thing around the Codec since there is one in `quarkus-langchain4j-core` but I have no idea on how to use it.

### CODE REVIEW

**Comment by geoand** [comment]
Very nice! I'll review this carefully next week.

**Comment by geoand** [comment]
I _think_  actually that the code could be a lot simpler, but I'll have to look closely and try stuff out next week.

**Comment by geoand** [comment]
I will push some changes to the PR shortly.

**Comment by geoand** [comment]
I pushed an update that now makes this PR complete. 

However I have some comments about naming I would like @jmartisk to comment on as well

File: chatmemorystore-redis/deployment/src/main/java/io/quarkiverse/langchain4j/chatmemorystore/redis/deployment/RedisChatMemoryStoreBuildTimeConfig.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import static io.quarkus.runtime.annotations.ConfigPhase.BUILD_TIME;
+
+import java.util.Optional;
+
+import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+
+@ConfigRoot(phase = BUILD_TIME)
+@ConfigMapping(prefix = "quarkus.langchain4j.chatmemorystore.redis")
```
Comment (geoand) [inline chatmemorystore-redis/deployment/src/main/java/io/quarkiverse/langchain4j/chatmemorystore/redis/deployment/RedisChatMemoryStoreBuildTimeConfig.java:11]: @jmartisk any suggestions on improving this name? Sounds like a little to much for users to type...

File: chatmemorystore-redis/deployment/src/main/java/io/quarkiverse/langchain4j/chatmemorystore/redis/deployment/RedisChatMemoryStoreProcessor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.redis.client.RedisClientName;
+import io.quarkus.redis.datasource.RedisDataSource;
+import io.quarkus.redis.deployment.client.RequestedRedisClientBuildItem;
+import io.quarkus.redis.runtime.client.config.RedisConfig;
+
+class RedisChatMemoryStoreProcessor {
+
+    public static final DotName REDIS_CHAT_MEMORY_STORE = DotName.createSimple(RedisChatMemoryStore.class);
+    private static final String FEATURE = "langchain4j-chatmemorystore-redis";
```
Comment (geoand) [inline chatmemorystore-redis/deployment/src/main/java/io/quarkiverse/langchain4j/chatmemorystore/redis/deployment/RedisChatMemoryStoreProcessor.java:28]: Same question about the naming goes here too

File: chatmemorystore-redis/runtime/pom.xml [hunk 3]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>io.quarkiverse.langchain4j</groupId>
+    <artifactId>quarkus-langchain4j-chatmemorystore-redis-parent</artifactId>
+    <version>999-SNAPSHOT</version>
+  </parent>
+  <artifactId>quarkus-langchain4j-chatmemorystore-redis</artifactId>
```
Comment (geoand) [inline chatmemorystore-redis/runtime/pom.xml:10]: Do we want this name or can we come up with something better?

**Comment by geoand** [comment]
🙏🏼

**Comment by jmartisk** [comment]
Looks nice!
I'm a little worried that now we will have two modules, `quarkus-langchain4j-redis` and `quarkus-langchain4j-chatmemorystore-redis`, it won't be very clear what the former is for. 
The solution going forward would be to either merge the two extensions (but then it might get too big), or to adopt a new naming strategy, where each extension contains in its name what it actually provides, so `quarkus-langchain4j-embedding-store-redis` and `quarkus-langchain4j-chat-memory-store-redis`, for example.
We should decide on a reasonable strategy before it gets too late I think

File: chatmemorystore-redis/runtime/pom.xml (hunk 3 above)
Comment (jmartisk) [reply chatmemorystore-redis/runtime/pom.xml:10]: I would use dashes in `chat-memory-store`, it is longer but `chatmemorystore` as one word looks weird. We don't use `embeddingstore` either.

**Comment by geoand** [comment]
> going forward would be to either merge the two extensions 

I don't think we should unless we really see in the real world that these two different concerns actually almost always met with the same Datastore.

> We should decide on a reasonable strategy before it gets too late I think

+1

**Comment by geoand** [comment]
@cescoffier do you have any good suggestions on the naming of the Redis modules?

**Comment by geoand** [comment]
Good points!

**Comment by geoand** [comment]
I am in favor of Clement's proposals, WDYT @jmartisk @sebastienblanc ?

**Comment by jmartisk** [comment]
So, let's sum it up:

- For AI model providers (chat model, embedding model, scoring model, image model, moderation model), we will clump all models provided from a provider into one extension and name it, say, `quarkus-langchain4j-PROVIDER`
- For embedding store providers, the extension will be named `quarkus-langchain4j-embedding-store-PROVIDER`
- For memory store providers, the extension will be named `quarkus-langchain4j-memory-store-PROVIDER`

Is this it?

> From the user POV, I believe quarkus-langchain4j-redis-memory-store is easier to read, and once you select redis, the auto-completion will give you all the possibilities.

I think I'd prefer the "developer" PoV, typic `memory-store` will also get you auto-completion for all memory stores - and there will be lots of them, while typing `redis` will only give you two possibilities - memory store and embedding store.

**Comment by sebastienblanc** [comment]
> I think I'd prefer the "developer" PoV, typing `memory-store` will also get you auto-completion for all memory stores - and there will be lots of them, while typing `redis` will only give you two possibilities - memory store and embedding store.
💯

**Comment by geoand** [comment]
> So, let's sum it up:
> 
> * For AI model providers (chat model, embedding model, scoring model, image model, moderation model), we will clump all models provided from a provider into one extension and name it, say, `quarkus-langchain4j-PROVIDER`
> * For embedding store providers, the extension will be named `quarkus-langchain4j-embedding-store-PROVIDER`
> * For memory store providers, the extension will be named `quarkus-langchain4j-memory-store-PROVIDER`
> 
> Is this it?
> 
> > From the user POV, I believe quarkus-langchain4j-redis-memory-store is easier to read, and once you select redis, the auto-completion will give you all the possibilities.
> 
> I think I'd prefer the "developer" PoV, typing `memory-store` will also get you auto-completion for all memory stores - and there will be lots of them, while typing `redis` will only give you two possibilities - memory store and embedding store.

@cescoffier do you agree with this proposal?

**Comment by sebastienblanc** [comment]
Should we decide on one of the naming conventions ?

**Comment by geoand** [comment]
cc @cescoffier

**Comment by sebastienblanc** [comment]
Just bumping this up ! (Really selfish agenda : I would like to release a
video/blogpost on this feature)

**Comment by geoand** [comment]
Clement is on PTO, so let's wait until Monday before moving forward

**Comment by geoand** [comment]
@cescoffier any preferences on the naming or should we move forward with this?

**Comment by sebastienblanc** [comment]
@geoand I've been messing up with my branches 😅 but yeah let me find some time today to try to fix it and do renaming.

**Comment by geoand** [comment]
🙏🏼

**Comment by sebastienblanc** [comment]
Renamed, Rebased and squashed commits ! Let me know if it is all good

**Review by jmartisk — APPROVED** [review]

File: memory-store-redis/deployment/src/test/java/io/quarkiverse/langchain4j/memorystore/redis/test/RedisChatMemoryStoreTest.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    @AfterAll
+    static void afterAll() {
+        wireMockServer.stop();
+    }
+
+    @BeforeEach
+    void setup() {
+        wireMockServer.resetAll();
+        wireMockServer.stubFor(WiremockUtils.defaultChatCompletionsStub(API_KEY));
+    }
```
Comment (geoand) [inline memory-store-redis/deployment/src/test/java/io/quarkiverse/langchain4j/memorystore/redis/test/RedisChatMemoryStoreTest.java:79]: Can we please use the utilities that are now in `quarkus-langchain4j-testing-internal`? It should make this code unnecessary

Comment (sebastienblanc) [reply memory-store-redis/deployment/src/test/java/io/quarkiverse/langchain4j/memorystore/redis/test/RedisChatMemoryStoreTest.java:79]: I'm trying this out right now but honestly I'm a bit lost

Comment (geoand) [reply memory-store-redis/deployment/src/test/java/io/quarkiverse/langchain4j/memorystore/redis/test/RedisChatMemoryStoreTest.java:79]: Let me know if you need any help

Comment (sebastienblanc) [reply memory-store-redis/deployment/src/test/java/io/quarkiverse/langchain4j/memorystore/redis/test/RedisChatMemoryStoreTest.java]: should be good now

**Comment by adriens** [comment]
:partying_face: :champagne:

**Pull request merged**